tests/core/snapd-refresh-vs-services-reboots/task.yaml

```diff
@@ -0,0 +1,137 @@
+  # end of the test to ensure that other tests use the snapd version that we
+  # started with, we revert the snapd snap to the original revision without 
+  # triggering a garbage collection of the revisions we reverted away from (i.e.
+  # the new revisions we installed as part of this test execution). This is 
+  # problematic because it leaves those old revisions mounted at /snap/snapd/x2
+  # for example and now at the end of the test the active revision of snapd is 
+  # x1. Then during the next test execution that tries to install a dangerous 
+  # local revision of snapd, it will have lost any reference to the previous x2
+  # revision (since we clear state.json at the end of the test execution in 
+  # reset.sh), and now snapd is trying to install and make active x2 again, and
+  # it will copy all the right files, but when it comes time to mount the new x2
+  # revision of snapd, there will already be an existing active mount unit for 
+  # the previous revision at /snap/snapd/x2, and so the bits of code that try
+  # to mount the .snap file there will effectively just silently fail since the
+  # is already a snapd snap mounted there, it is just the wrong one mounted 
+  # there. 
+
+  # We remedy this here, temporarily, by first manually executing all cleanups
+  # that were deferred, since one of those deferred cleanups is likely the 
+  # revert to the previous revision of the snapd snap, and then we manually 
+  # remove all disabled revisions of the snapd snap - this manual removal will
+  # in fact unmount the mount units for /snap/snapd/x2 for example, avoiding the
+  # bug.
+
+  tests.cleanup restore
+  for rev in $(snap list snapd --all | grep disabled | awk '{print $3}'); do
+    snap remove snapd --revision="$rev"
+  done
+
+execute: |
+  if ! os.query is-pc-amd64 && ! os.query is-arm; then
+    echo "architecture not supported for this variant"
+    exit 0
+  fi
+
+  if [ "$SPREAD_REBOOT" = "0" ]; then
+    # install the special service snap which will fail to startup every other time
+    # it is started
+    snap pack test-snapd-svc-flip-flop --filename=app.snap
+
+    # install the snap so it is running
+    snap install --dangerous app.snap
+
+    # ensure that the service is active and has a running PID
+    snap services | MATCH 'test-snapd-svc-flip-flop\.svc1\s+enabled\s+active.*'
+
+    # ensure that the service was installed with Requires= which triggers the bug
+    MATCH Requires=usr-lib-snapd.mount < /etc/systemd/system/snap.test-snapd-svc-flip-flop.svc1.service
+
+    # mock the reboot command so that when we try to refresh snapd and it fails,
+    # we can observe this and reboot with spread instead
+
+    # TODO: make this a common helper for mocking expected snapd reboots, this
+    # code is shared with uc20-recovery.sh right now
+
+    # save the original systemctl command since we essentially need to mock it
+    cp /bin/systemctl /tmp/orig-systemctl
+    chmod +x /tmp/orig-systemctl
```

Comment:
This should not be needed, as `cp` preserves the executable permission

Reply:
hmm, in my testing I found it necessary, but it has been a while since I originally worked on the test

Reply:
you are right it is not necessary, thanks for the pointer